README.md
# Logback GELF

[![Build Status](https://api.travis-ci.org/osiegmar/logback-gelf.svg)](https://travis-ci.org/osiegmar/logback-gelf)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/de.siegmar/logback-gelf/badge.svg)](https://maven-badges.herokuapp.com/maven-central/de.siegmar/logback-gelf)

Logback appender for sending GELF (Graylog Extended Log Format) messages with zero additional
dependencies.


## Latest release

The most recent release is 2.2.0, released December 14, 2019.

To add a dependency using Maven, use the following:

```xml
<dependency>
    <groupId>de.siegmar</groupId>
    <artifactId>logback-gelf</artifactId>
    <version>2.2.0</version>
    <scope>runtime</scope>
</dependency>
```

To add a dependency using Gradle:

```gradle
dependencies {
    runtimeOnly 'de.siegmar:logback-gelf:2.2.0'
}
```


## Features

- UDP (with chunking)
- TCP (with or without TLS encryption)
- Deflate compression in UDP mode
- Client side load balancing (round robin)
- Forwarding of MDC (Mapped Diagnostic Context)
- Forwarding of caller data
- Forwarding of static fields
- Forwarding of exception root cause
- No runtime dependencies beside Logback


## Requirements

- Java 8
- Logback 1.2.3


## Important notes

Some changes may require to update your configuration.

### Breaking changes in version 3
* Version 3.0.0 of this library upgraded from Java 7 to Java 8.
* The server's certificate hostname now gets verified by `GelfTcpTlsAppender`.
* The `trustAllCertificates` property of `GelfTcpTlsAppender` was renamed to `insecure`.

### Breaking changes in version 2
* Version 2.0.0 of this library introduced a configuration change.

**Old** format:
```xml
<layout class="de.siegmar.logbackgelf.GelfLayout">
```

**New** format:
```xml
<encoder class="de.siegmar.logbackgelf.GelfEncoder">
```

This change was introduced, as the API of the Encoder interface changed in Logback 1.2.0.


## Examples

Simple UDP configuration:

```xml
<configuration>

    <appender name="GELF" class="de.siegmar.logbackgelf.GelfUdpAppender">
        <graylogHost>localhost</graylogHost>
        <graylogPort>12201</graylogPort>
    </appender>

    <root level="debug">
        <appender-ref ref="GELF" />
    </root>

</configuration>
```

Simple TCP configuration:

```xml
<configuration>

    <appender name="GELF" class="de.siegmar.logbackgelf.GelfTcpAppender">
        <graylogHost>localhost</graylogHost>
        <graylogPort>12201</graylogPort>
    </appender>

    <!-- Use AsyncAppender to prevent slowdowns -->
    <appender name="ASYNC GELF" class="ch.qos.logback.classic.AsyncAppender">
        <appender-ref ref="GELF" />
    </appender>

    <root level="debug">
        <appender-ref ref="ASYNC GELF" />
    </root>

</configuration>
```

Simple TCP with TLS configuration:

```xml
<configuration>

    <appender name="GELF" class="de.siegmar.logbackgelf.GelfTcpTlsAppender">
        <graylogHost>localhost</graylogHost>
        <graylogPort>12201</graylogPort>
    </appender>

    <!-- Use AsyncAppender to prevent slowdowns -->
    <appender name="ASYNC GELF" class="ch.qos.logback.classic.AsyncAppender">
        <appender-ref ref="GELF" />
    </appender>

    <root level="debug">
        <appender-ref ref="ASYNC GELF" />
    </root>

</configuration>
```

Find more advanced examples in the [examples directory](examples).


## Configuration

### Appender

`de.siegmar.logbackgelf.GelfUdpAppender`

* **graylogHost**: IP or hostname of graylog server.
  If the hostname resolves to multiple ip addresses, round robin will be used.
* **graylogPort**: Port of graylog server. Default: 12201.
* **encoder**: See Encoder configuration below.
* **maxChunkSize**: Maximum size of GELF chunks in bytes. Default chunk size is 508 - this prevents
  IP packet fragmentation. This is also the recommended minimum.
  Maximum supported chunk size is 65,467 bytes.
* **useCompression**: If true, compression of GELF messages is enabled. Default: true.


`de.siegmar.logbackgelf.GelfTcpAppender`

* **graylogHost**: IP or hostname of graylog server.
  If the hostname resolves to multiple ip addresses, round robin will be used.
* **graylogPort**: Port of graylog server. Default: 12201.
* **encoder**: See Encoder configuration below.
* **connectTimeout**: Maximum time (in milliseconds) to wait for establishing a connection. A value
  of 0 disables the connect timeout. Default: 15,000 milliseconds.
* **reconnectInterval**: Time interval (in seconds) after an existing connection is closed and
  re-opened. A value of -1 disables automatic reconnects. Default: 60 seconds.
* **maxRetries**: Number of retries. A value of 0 disables retry attempts. Default: 2.
* **retryDelay**: Time (in milliseconds) between retry attempts. Ignored if maxRetries is 0.
  Default: 3,000 milliseconds.
* **poolSize**: Number of concurrent tcp connections (minimum 1). Default: 2.
* **poolMaxWaitTime**: Maximum amount of time (in milliseconds) to wait for a connection to become
  available from the pool. A value of -1 disables the timeout. Default: 5,000 milliseconds.


`de.siegmar.logbackgelf.GelfTcpTlsAppender`

* Everything from GelfTcpAppender
* **trustedServerCertificate**: An optionally configured X.509 server certificate (PEM encoded) to
  trust (whitelist). Can be configured multiple times to ease certification renewal.
  If configured, the server needs to offer one of these certificates in order to allow
  communication. The certificate offered by the server needs to be valid (matching hostname and
  not expired). If this property is configured, the server's certificate chain is not validated in
  order to allow self-signed certificates. Default: none.
* **insecure**: If true, skip the TLS certificate validation.
  You should not use this in production! Default: false.


### Encoder

`de.siegmar.logbackgelf.GelfEncoder`

* **originHost**: Origin hostname - will be auto detected if not specified.
* **includeRawMessage**: If true, the raw message (with argument placeholders) will be sent, too.
  Default: false.
* **includeMarker**: If true, logback markers will be sent, too. Default: true.
* **includeMdcData**: If true, MDC keys/values will be sent, too. Default: true.
* **includeCallerData**: If true, caller data (source file-, method-, class name and line) will be
  sent, too. Default: false.
* **includeRootCauseData**: If true, root cause exception of the exception passed with the log
   message will be exposed in the root_cause_class_name and root_cause_message fields.
   Default: false.
* **includeLevelName**: If true, the log level name (e.g. DEBUG) will be sent, too. Default: false.
* **levelNameKey**: The key (i.e. the field name) that should be used for the log level name. 
  This is only relevant when includeLevelName is true. Default: level_name.
* **loggerNameKey**: The key (i.e. the field name) that should be used for the logger name. 
  Default: logger_name.
* **threadNameKey**: The key (i.e. the field name) that should be used for the thread name. 
  Default: thread_name.
* **appendNewline**: If true, a system depended newline separator will be added at the end of each message.
  Don't use this in conjunction with TCP or UDP appenders, as this is only reasonable for
  console logging!
* **shortPatternLayout**: Short message format. Default: `"%m%nopex"`.
* **fullPatternLayout**: Full message format (Stacktrace). Default: `"%m%n"`.
* **numbersAsString**: Log numbers as String. Default: true (will be changed in next major release).
* **staticFields**: Additional, static fields to send to graylog. Defaults: none.


## Troubleshooting

If you have any problems, enable the debug mode and check the logs.

```xml
<configuration debug="true">
    ...
</configuration>
```


## Contribution

- Fork
- Code
- Add test(s)
- Commit
- Send me a pull request


## Copyright

Copyright (C) 2016-2019 Oliver Siegmar

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.0.0] - 2020-??-??
### Added
- Allow encoder subclasses to customize the message before it is converted to String
  [\#40](https://github.com/osiegmar/logback-gelf/issues/40)
- Server certificate hostname verification in `GelfTcpTlsAppender`.
- Add server certificate whitelisting (`trustedServerCertificate`).

### Changed
- Upgrade to Java 8 (Premier Support of Java 7 ended in July 2019)
- Rename `trustAllCertificates` property of `GelfTcpTlsAppender` to `insecure`.

## [2.2.0] - 2019-12-14
### Added
- Add customizable keys for the logger name and thread name (#41)
  [\#41](https://github.com/osiegmar/logback-gelf/issues/41)

### Fixed
- Fix handling of multiple markers
  [\#35](https://github.com/osiegmar/logback-gelf/issues/35)

## [2.1.2] - 2019-11-04
### Fixed
- Build needs to be performed with Java < 9
  [\#38](https://github.com/osiegmar/logback-gelf/issues/38)

## [2.1.1] - 2019-11-03
### Fixed
- Fix empty hostname
  [\#34](https://github.com/osiegmar/logback-gelf/issues/34)

## [2.1.0] - 2019-06-12
### Changed
- Log numeric values as number (double precision) not string
  [\#30](https://github.com/osiegmar/logback-gelf/pull/30)

## [2.0.1] - 2019-05-19
### Fixed
- Reopen the UDP channel, if it was closed
  [\#20](https://github.com/osiegmar/logback-gelf/issues/20)

## [2.0.0] - 2019-02-12
### Changed
- Update dependency to logback 1.2.3
  [\#21](https://github.com/osiegmar/logback-gelf/issues/21)
- Changed implementation from Layout to Encoder (also renamed class GelfLayout to GelfEncoder and layout to encoder in GelfAppender)

## [1.1.0] - 2018-01-21
### Added
- Simple connection pooling in GelfTcpAppender & round robin host lookup in GelfUdpAppender
  [\#11](https://github.com/osiegmar/logback-gelf/issues/11)

### Changed
- Update dependency to logback 1.1.8
  [\#1](https://github.com/osiegmar/logback-gelf/issues/1)

### Fixed
- The reconnect interval could not be disabled
  [\#12](https://github.com/osiegmar/logback-gelf/issues/12)

## [1.0.4] - 2017-04-03
### Added
- Support for GELF console logging (appendNewline in GelfLayout)

### Fixed
- Fix interrupted flag in GelfTcpAppender (restore flag after catching InterruptedException)

## [1.0.3] - 2017-01-12
### Added
- Support for logback 1.1.8
  [\#6](https://github.com/osiegmar/logback-gelf/issues/6)

## [1.0.2] - 2016-10-18
### Fixed
- Fix possible infinite loop bug with exception root cause

## [1.0.1] - 2016-10-08
### Added
- Support for forwarding of exception root cause
- Support for TLS encryption (with TCP)

## 1.0.0 - 2016-03-20

- Initial release

[Unreleased]: https://github.com/osiegmar/logback-gelf/compare/v2.2.0...HEAD
[2.2.0]: https://github.com/osiegmar/logback-gelf/compare/v2.1.2...v2.2.0
[2.1.2]: https://github.com/osiegmar/logback-gelf/compare/v2.1.1...v2.1.2
[2.1.1]: https://github.com/osiegmar/logback-gelf/compare/v2.1.0...v2.1.1
[2.1.0]: https://github.com/osiegmar/logback-gelf/compare/v2.0.1...v2.1.0
[2.0.1]: https://github.com/osiegmar/logback-gelf/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/osiegmar/logback-gelf/compare/v1.1.0...v2.0.0
[1.1.0]: https://github.com/osiegmar/logback-gelf/compare/v1.0.4...v1.1.0
[1.0.4]: https://github.com/osiegmar/logback-gelf/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/osiegmar/logback-gelf/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/osiegmar/logback-gelf/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/osiegmar/logback-gelf/compare/v1.0.0...v1.0.1


src/test/java/de/siegmar/logbackgelf/GelfTcpTlsAppenderTest.java
/*
 * Logback GELF - zero dependencies Logback GELF appender library.
 * Copyright (C) 2016 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackgelf;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.DataInputStream;
import java.io.IOException;
import java.net.Socket;
import java.security.cert.X509Certificate;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import javax.net.ssl.SSLServerSocket;
import javax.net.ssl.SSLServerSocketFactory;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.io.ByteStreams;

import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.LoggerContext;
import ch.qos.logback.core.status.Status;

public class GelfTcpTlsAppenderTest {

    private static final String LOGGER_NAME = GelfTcpTlsAppenderTest.class.getCanonicalName();

    private TcpServerRunnable server;

    public GelfTcpTlsAppenderTest() {
        final String mySrvKeystore =
            GelfTcpTlsAppenderTest.class.getResource("/mySrvKeystore").getFile();
        System.setProperty("javax.net.ssl.keyStore", mySrvKeystore);
        System.setProperty("javax.net.ssl.keyStorePassword", "secret");
    }

    @BeforeEach
    public void before() throws IOException, InterruptedException {
        server = new TcpServerRunnable();
        final Thread serverThread = new Thread(server);
        serverThread.start();
    }

    @Test
    public void configurationError() throws Exception {
        final LoggerContext lc = (LoggerContext) LoggerFactory.getILoggerFactory();

        final AtomicReference<Status> lastStatus = new AtomicReference<>();
        lc.getStatusManager().add(lastStatus::set);

        final GelfEncoder gelfEncoder = new GelfEncoder();
        gelfEncoder.setContext(lc);
        gelfEncoder.setOriginHost("localhost");
        gelfEncoder.start();

        final Logger logger = (Logger) LoggerFactory.getLogger(LOGGER_NAME);
        logger.addAppender(buildAppender(lc, gelfEncoder));
        logger.setAdditive(false);

        final X509Certificate cert = new X509Util.CertBuilder()
            .build(null, "foo.example.com");

        final GelfTcpTlsAppender gelfAppender = new GelfTcpTlsAppender();
        gelfAppender.setContext(lc);
        gelfAppender.setName("GELF");
        gelfAppender.setEncoder(gelfEncoder);
        gelfAppender.setGraylogHost("localhost");
        gelfAppender.setGraylogPort(server.getPort());
        gelfAppender.setInsecure(true);
        gelfAppender.addTrustedServerCertificate(X509Util.toPEM(cert));
        gelfAppender.start();

        final IllegalStateException e = (IllegalStateException) lastStatus.get().getThrowable();
        assertEquals("Configuration options 'insecure' and 'trustedServerCertificates' "
            + "are mutually exclusive!", e.getMessage());
    }

    @Test
    public void simple() throws IOException {
        final Logger logger = setupLogger();

        logger.error("Test message");

        stopLogger(logger);

        final JsonNode jsonNode = receiveMessage();
        assertEquals("1.1", jsonNode.get("version").textValue());
        assertEquals("localhost", jsonNode.get("host").textValue());
        assertEquals("Test message", jsonNode.get("short_message").textValue());
        assertTrue(jsonNode.get("timestamp").isNumber());
        assertEquals(3, jsonNode.get("level").intValue());
        assertNotNull(jsonNode.get("_thread_name").textValue());
        assertEquals(LOGGER_NAME, jsonNode.get("_logger_name").textValue());
    }

    private Logger setupLogger() {
        final LoggerContext lc = (LoggerContext) LoggerFactory.getILoggerFactory();

        final GelfEncoder gelfEncoder = new GelfEncoder();
        gelfEncoder.setContext(lc);
        gelfEncoder.setOriginHost("localhost");
        gelfEncoder.start();

        final Logger logger = (Logger) LoggerFactory.getLogger(LOGGER_NAME);
        logger.addAppender(buildAppender(lc, gelfEncoder));
        logger.setAdditive(false);

        return logger;
    }

    private GelfTcpTlsAppender buildAppender(final LoggerContext lc, final GelfEncoder encoder) {
        final GelfTcpTlsAppender gelfAppender = new GelfTcpTlsAppender();
        gelfAppender.setContext(lc);
        gelfAppender.setName("GELF");
        gelfAppender.setEncoder(encoder);
        gelfAppender.setGraylogHost("localhost");
        gelfAppender.setGraylogPort(server.getPort());
        gelfAppender.setInsecure(true);
        gelfAppender.start();
        return gelfAppender;
    }

    private JsonNode receiveMessage() throws IOException {
        return new ObjectMapper().readTree(server.getReceivedData());
    }

    private void stopLogger(final Logger logger) {
        final GelfTcpTlsAppender gelfAppender = (GelfTcpTlsAppender) logger.getAppender("GELF");
        gelfAppender.stop();
    }

    private static final class TcpServerRunnable implements Runnable {

        private final SSLServerSocket server;
        private final Semaphore semaphore = new Semaphore(1);
        private byte[] receivedData;

        TcpServerRunnable() throws IOException, InterruptedException {
            final SSLServerSocketFactory sslserversocketfactory =
                (SSLServerSocketFactory) SSLServerSocketFactory.getDefault();
            server = (SSLServerSocket) sslserversocketfactory.createServerSocket(0);
            semaphore.acquire();
        }

        int getPort() {
            return server.getLocalPort();
        }

        byte[] getReceivedData() {
            try {
                if (!semaphore.tryAcquire(10, TimeUnit.SECONDS)) {
                    throw new IllegalStateException("Couldn't acquire semaphore!");
                }
            } catch (final InterruptedException e) {
                throw new IllegalStateException(e);
            }
            return receivedData;
        }

        @Override
        public void run() {
            try (Socket socket = server.accept()) {
                try (DataInputStream in = new DataInputStream(socket.getInputStream())) {
                    receivedData = ByteStreams.toByteArray(in);
                }
            } catch (final IOException e) {
                throw new IllegalStateException(e);
            }

            try {
                server.close();
            } catch (final IOException e) {
                throw new IllegalStateException(e);
            }

            semaphore.release();
        }

    }

}


src/test/java/de/siegmar/logbackgelf/CustomX509TrustManagerTest.java
/*
 * Logback GELF - zero dependencies Logback GELF appender library.
 * Copyright (C) 2019 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackgelf;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.io.IOException;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.time.LocalDate;
import java.util.Collections;

import javax.net.ssl.TrustManager;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class CustomX509TrustManagerTest {

    private static final String HOSTNAME = "graylog.foo.bar";

    private CustomX509TrustManager tm = new CustomX509TrustManager(
        defaultTrustManager(null), HOSTNAME, Collections.emptyList());
    private X509Util.CertBuilder c;

    @BeforeEach
    public void before() throws Exception {
        c = new X509Util.CertBuilder();
    }

    @SuppressWarnings("checkstyle:illegalcatch")
    @Test
    public void selfSigned() {
        // The default trust manager will reject a self-signed certificate
        try {
            validate(c.build(null, HOSTNAME));
            fail();
        } catch (Exception e) {
            assertTrue(e.getMessage().contains("unable to find valid certification path"));
        }
    }

    @Test
    public void selfSignedInsecure() throws Exception {
        tm = new CustomX509TrustManager(noopTrustManager(), HOSTNAME, Collections.emptyList());
        validate(c.build(null, HOSTNAME));
    }

    @Test
    public void selfSignedTrusted() throws Exception {
        final X509Certificate certificate = c.build(null, HOSTNAME);

        tm = new CustomX509TrustManager(defaultTrustManager(null), HOSTNAME,
            Collections.singletonList(certificate));

        validate(certificate);
    }

    @Test
    public void validCert() throws Exception {
        tm = new CustomX509TrustManager(noopTrustManager(), HOSTNAME, Collections.emptyList());

        validate(c.build(null, HOSTNAME));
        validate(c.build(null, "x.foo.bar", HOSTNAME));
        validate(c.build(null, "x.foo.bar", "*.foo.bar"));
        validate(c.build(HOSTNAME));
    }

    @Test
    public void nameMismatch() {
        tm = new CustomX509TrustManager(noopTrustManager(), HOSTNAME, Collections.emptyList());

        final String wrongName = "graylog2.foo.bar";

        CertificateException e = assertThrows(CertificateException.class,
            () -> validate(c.build(null, wrongName)));
        assertEquals("Server certificate mismatch. Tried to verify graylog.foo.bar "
            + "against subject alternative names: [graylog2.foo.bar]", e.getMessage());

        e = assertThrows(CertificateException.class, () -> validate(c.build(wrongName)));
        assertEquals("Server certificate mismatch. Tried to verify graylog.foo.bar "
            + "against common name: graylog2.foo.bar", e.getMessage());
    }

    @Test
    public void caSigned() throws Exception {
        final X509Util.CABuilder caBuilder = new X509Util.CABuilder();
        final X509Certificate cert = prepareCaSigned(caBuilder)
            .build(HOSTNAME);

        validate(cert, caBuilder.getCaCertificate());
    }

    private X509Util.CertBuilder prepareCaSigned(final X509Util.CABuilder caBuilder)
        throws KeyStoreException, IOException, NoSuchAlgorithmException, CertificateException {

        final KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
        ks.load(null, null);
        ks.setCertificateEntry("my-cert", caBuilder.getCaCertificate());

        tm = new CustomX509TrustManager(defaultTrustManager(ks), HOSTNAME, Collections.emptyList());

        return c
            .caPrivateKey(caBuilder.getKeyPair().getPrivate())
            .caCertificate(caBuilder.getCaCertificate());
    }

    @Test
    public void expired() throws Exception {
        final X509Util.CABuilder caBuilder = new X509Util.CABuilder();
        final X509Certificate cert = prepareCaSigned(caBuilder)
            .validFrom(LocalDate.now().minusYears(2))
            .validTo(LocalDate.now().minusYears(1))
            .build(HOSTNAME);

        assertThrows(CertificateException.class,
            () -> validate(cert, caBuilder.getCaCertificate()));
    }

    @Test
    public void caSignedNotWhitelisted() throws Exception {
        final X509Util.CABuilder caBuilder = new X509Util.CABuilder();
        final X509Certificate cert = prepareCaSigned(caBuilder)
            .build(HOSTNAME);

        tm = new CustomX509TrustManager(defaultTrustManager(null), HOSTNAME,
            Collections.singletonList(c.build(null, HOSTNAME)));

        final CertificateException e = assertThrows(CertificateException.class,
            () -> validate(cert, caBuilder.getCaCertificate()));

        assertEquals("Server did not offer a whitelisted certificate", e.getMessage());
    }

    private void validate(final X509Certificate... certificates) throws CertificateException {
        tm.checkServerTrusted(certificates, "RSA");
    }

    private static X509TrustManager defaultTrustManager(final KeyStore keyStore) {
        try {
            final TrustManagerFactory trustManagerFactory =
                TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
            trustManagerFactory.init(keyStore);

            for (final TrustManager trustManager : trustManagerFactory.getTrustManagers()) {
                if (trustManager instanceof X509TrustManager) {
                    return (X509TrustManager) trustManager;
                }
            }

            return null;
        } catch (NoSuchAlgorithmException | KeyStoreException e) {
            throw new IllegalStateException(e);
        }
    }

    private static NoopX509TrustManager noopTrustManager() {
        return new NoopX509TrustManager();
    }

}


src/main/java/de/siegmar/logbackgelf/EasyX509TrustManager.java
/*
 * Logback GELF - zero dependencies Logback GELF appender library.
 * Copyright (C) 2019 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackgelf;

import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.List;

import javax.net.ssl.X509TrustManager;

public class EasyX509TrustManager implements X509TrustManager {

    private static final int IDX_KEY_CERT_SIGN = 5;
    private List<X509Certificate> trustedServerCertificates;

    public List<X509Certificate> getTrustedServerCertificates() {
        return trustedServerCertificates;
    }

    public void setTrustedServerCertificates(final List<X509Certificate> certificates) {
        for (final X509Certificate trustedServerCertificate : certificates) {
            if (isCaCertificate(trustedServerCertificate)) {
                throw new IllegalArgumentException("Not a server certificate");
            }
        }
        this.trustedServerCertificates = certificates;
    }

    @Override
    public void checkClientTrusted(final X509Certificate[] chain, final String authType) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void checkServerTrusted(final X509Certificate[] chain, final String authType)
        throws CertificateException {

        if (trustedServerCertificates != null && !trustedServerCertificates.isEmpty()) {
            validateServerCertificate(chain);
        }
    }

    private void validateServerCertificate(final X509Certificate[] chain)
        throws CertificateException {

        for (final X509Certificate cert : chain) {
            if (isCaCertificate(cert)) {
                // CA certificate
                continue;
            }

            for (final X509Certificate trustedServerCertificate : trustedServerCertificates) {
                if (cert.equals(trustedServerCertificate)) {
                    cert.checkValidity();
                    return;
                }
            }
        }

        throw new CertificateException("Server certificate mismatch");
    }

    private static boolean isCaCertificate(final X509Certificate cert) {
        final boolean[] keyUsage = cert.getKeyUsage();
        return keyUsage != null
            && keyUsage.length > IDX_KEY_CERT_SIGN
            && keyUsage[IDX_KEY_CERT_SIGN];
    }

    @Override
    public X509Certificate[] getAcceptedIssuers() {
        return new X509Certificate[0];
    }

}


src/main/java/de/siegmar/logbackgelf/CustomX509TrustManager.java
/*
 * Logback GELF - zero dependencies Logback GELF appender library.
 * Copyright (C) 2019 Oliver Siegmar
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package de.siegmar.logbackgelf;

import java.security.cert.CertificateException;
import java.security.cert.CertificateParsingException;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import javax.naming.InvalidNameException;
import javax.naming.ldap.LdapName;
import javax.naming.ldap.Rdn;
import javax.net.ssl.X509TrustManager;

class CustomX509TrustManager implements X509TrustManager {

    private static final int TYPE_DNS_NAME = 2;

    private final X509TrustManager defaultTrustManager;
    private final String hostname;
    private final List<X509Certificate> trustedServerCertificates;

    CustomX509TrustManager(final X509TrustManager defaultTrustManager, final String hostname,
                           final List<X509Certificate> trustedServerCertificates) {
        this.defaultTrustManager = Objects.requireNonNull(defaultTrustManager);
        this.hostname = Objects.requireNonNull(hostname);
        this.trustedServerCertificates = Objects.requireNonNull(trustedServerCertificates);
    }

    @Override
    public void checkClientTrusted(final X509Certificate[] chain, final String authType) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void checkServerTrusted(final X509Certificate[] chain, final String authType)
        throws CertificateException {

        // The first certificate is the server certificate
        final X509Certificate serverCert = chain[0];

        if (trustedServerCertificates.isEmpty()) {
            // If not explicitly trusted, check via the default trust manager (chain validation)
            defaultTrustManager.checkServerTrusted(chain, authType);
        } else {
            if (!trustedServerCertificates.contains(serverCert)) {
                throw new CertificateException("Server did not offer a whitelisted certificate");
            }

            // Check if the certificate is valid. This is also done by the default trust manager.
            serverCert.checkValidity();
        }

        if (checkAlternativeNames(serverCert)) {
            return;
        }

        // Check the deprecated common name
        checkCommonName(serverCert);
    }

    private boolean checkAlternativeNames(final X509Certificate serverCert)
        throws CertificateException {

        final List<String> alternativeNames = getAlternativeNames(serverCert);
        if (!alternativeNames.isEmpty()) {
            for (final String alternativeName : alternativeNames) {
                if (HostnameVerifier.verify(hostname, alternativeName)) {
                    return true;
                }
            }

            throw new CertificateException(String.format("Server certificate mismatch. Tried to "
                + "verify %s against subject alternative names: %s", hostname, alternativeNames));
        }
        return false;
    }

    private static List<String> getAlternativeNames(final X509Certificate cert)
        throws CertificateParsingException {

        final Collection<List<?>> subjectAlternativeNames = cert.getSubjectAlternativeNames();
        if (subjectAlternativeNames == null) {
            return Collections.emptyList();
        }

        final List<String> ret = new ArrayList<>();
        for (final List<?> entry : subjectAlternativeNames) {
            if (((Integer) entry.get(0)) == TYPE_DNS_NAME) {
                final String dNSName = (String) entry.get(1);
                if (dNSName != null) {
                    ret.add(dNSName);
                }
            }
        }

        return ret;
    }

    private void checkCommonName(final X509Certificate serverCert) throws CertificateException {
        final String commonName;
        try {
            commonName = getCommonName(serverCert);
            if (HostnameVerifier.verify(hostname, commonName)) {
                return;
            }
        } catch (final InvalidNameException e) {
            throw new CertificateException("Could not read CN from certificate", e);
        }

        throw new CertificateException(String.format("Server certificate mismatch. "
            + "Tried to verify %s against common name: %s", hostname, commonName));
    }

    private static String getCommonName(final X509Certificate cert) throws InvalidNameException {
        final LdapName ldapName = new LdapName(cert.getSubjectDN().getName());
        for (final Rdn rdn : ldapName.getRdns()) {
            if ("CN".equalsIgnoreCase(rdn.getType())) {
                return (String) rdn.getValue();
            }
        }

        throw new InvalidNameException("No common name found");
    }

    @Override
    public X509Certificate[] getAcceptedIssuers() {
        return defaultTrustManager.getAcceptedIssuers();
    }

}
